single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,751 @@
+        licenses.
+        """
+        licenses = [
+            "snap",
+            "mongodb-exporter",
+            "percona-backup-mongodb",
+            "percona-server",
+        ]
+        prefix = Path("./src/licenses") if self.substrate == Substrates.VM else Path("./")
+        # Create the directory if needed.
+        if self.substrate == Substrates.VM:
+            prefix.mkdir(exist_ok=True)
+            file = Path("./LICENSE")
+            dst = prefix / "LICENSE-charm"
+            self.workload.copy_to_unit(file, dst)
+        else:
+            name = "LICENSE-rock"
+            file = Path(f"{self.workload.paths.licenses_path}/{name}")
+            dst = prefix / name
+            if not dst.is_file():
+                self.workload.copy_to_unit(file, dst)
+
+        for license in licenses:
+            name = f"LICENSE-{license}"
+            file = Path(f"{self.workload.paths.licenses_path}/{name}")
+            dst = prefix / name
+            if not dst.is_file():
+                self.workload.copy_to_unit(file, dst)
+
+    def set_permissions(self) -> None:
+        """Ensure directories and make permissions.
+
+        We must ensure that the log status directory for LogRotate is existing.
+        We must also ensure that all data, log and log status directories have
+        the correct permissions.
+        """
+        self.workload.mkdir(LogRotateConfig.log_status_dir, make_parents=True)
+
+        for path in (
+            self.workload.paths.data_path,
+            self.workload.paths.logs_path,
+            LogRotateConfig.log_status_dir,
+        ):
+            self.workload.exec(
+                [
+                    "chown",
+                    "-R",
+                    f"{self.workload.users.user}:{self.workload.users.group}",
+                    f"{path}",
+                ]
+            )
+
+    def _initialise_replica_set(self):
+        """Helpful method to initialise the replica set and the users.
+
+        This is executed only by the leader.
+        This function first initialises the replica set, and then the three charm users.
+        Finally, if there are any integrated clients (direct clients in the
+        case of replication, or mongos clients in case of config-server),
+        oversee the relation to create the associated users.
+        At the very end, it sets the `db_initialised` flag to True.
+        """
+        if not self.model.unit.is_leader():
+            return
+        self.mongo_manager.initialise_replica_set()
+        self.mongo_manager.initialise_users()
+        logger.info("Manage client relation users")
+        if self.state.is_role(MongoDBRoles.REPLICATION):
+            for relation in self.state.client_relations:
+                self.mongo_manager.oversee_relation(relation)
+        elif self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            for relation in self.state.cluster_relations:
+                self.mongo_manager.oversee_relation(relation)
```

Comment:
I'd rather keep it as is because sometimes too much density loses readability and leads to painful bugs like we had with this exact piece of code in `mongodb-operator` charm.

Here, we know exactly who we are and which relation we are allowed to act upon. This prevents from accidentally modifying relations we're not supposed to change because of our role.